test/key_test.py

```diff
@@ -24,19 +24,24 @@ def test_import(self):
         "does it import"
         import pygame.key
 
-    def todo_test_get_focused(self):
-
-        # __doc__ (as of 2008-08-02) for pygame.key.get_focused:
-
-        # pygame.key.get_focused(): return bool
-        # true if the display is receiving keyboard input from the system
-        #
-        # This is true when the display window has keyboard focus from the
-        # system. If the display needs to ensure it does not lose keyboard
-        # focus, it can use pygame.event.set_grab() to grab all input.
-        #
-
-        self.fail()
+    def test_get_focused(self):
+        focused = pygame.key.get_focused()
+        if not SDL1:
+            self.assertFalse(focused) #No window to focus
+        self.assertIsInstance(focused, int)
+        display_sizes = pygame.display.list_modes()
+        if display_sizes == -1:
+            display_sizes = [(500, 500)]
+        if not display_sizes:
+            return
+        pygame.display.set_mode(size = display_sizes[-1], flags = pygame.FULLSCREEN)
+        pygame.event.set_grab(True)
+        pygame.event.pump() #Pump event queue to get window focus on macos
+        if not pygame.event.get_grab():
+            return
+        focused = pygame.key.get_focused()
+        self.assertIsInstance(focused, int)
+        self.assertTrue(focused)
```

Comment:
Just getting back to pygame after a summer holiday, apologies for the long wait.

My overall feedback is that this is a fine test for getting and testing a keyboard focus positive, but to fully test it is all working we'd want to try something we would expect would make the window _lose_ keyboard focus (i.e. something like minimising the window with `display.iconify()` - though you will want to check the test for that function to see what shenanigans are required for the iconify() to be registered by the OS and report back to pygame).

Complicating all of this (as with many unit tests) some of the window and display functionality does not work correctly on the 'dummy' display driver used on the continuous integration machines, I would not be at all suprised if you need to split this test into two functions - one, like your test above, that works on the CI - and one that has more testing in it but will only work on 'local' machines that use a real display driver.

You can mimic the 'dummy' test environment locally by setting the display driver used with something like:

```
import os

os.environ['SDL_VIDEODRIVER'] = 'dummy'
import pygame
````

(more info  here: https://www.pygame.org/wiki/DummyVideoDriver)

Just going to whiz through my inbox now but I may pop back to this later depending on time.

Reply:
I have changed the test to include iconifying the display. I also tested using SDL 1 and using the dummy display. I see get_focused() always returns true when using SDL 1. When using SDL 2 and a dummy display, get_focused always returns false. I have rewritten the test to sidestep most of the assertions when using a dummy display driver, but I have allowed the test to continue when using SDL 1, as always returning 1 seems like it is behaving incorrectly to me. Let me know if I should make more changes.